### PR TITLE
Fix invalid preferences

### LINKED ISSUE #296: Unable to Plugin to \Magento\TwoFactorAuth\Observer\ControllerActionPredispatch
### Preconditions (*)
1. Magento 2.4 

### Steps to reproduce (*)
1. Attempt to plugin to `\Magento\TwoFactorAuth\Observer\ControllerActionPredispatch::execute` will not work due to \Magento\TwoFactorAuth\TestFramework\ControllerActionPredispatch being a preference of it
2. Attempt to plugin to \Magento\TwoFactorAuth\TestFramework\ControllerActionPredispatch will work in dev mode as interceptors are generated on the fly
3. Switch to production mode and run `setup:di:compile`
4. No interceptor is generated for `\Magento\TwoFactorAuth\TestFramework\ControllerActionPredispatch` becuase it has `<VENDOR_MODULE>/Test` in its filepath (excluded in `\Magento\Setup\Module\Di\Code\Reader\ClassesScanner::extract`)

### Expected result (*)
1. should be able to plugin to `\Magento\TwoFactorAuth\Observer\ControllerActionPredispatch::execute`

### Actual result (*)
1. Unable to plugin to method due to preference

### Suggested solution
- Migrate the `\Magento\TwoFactorAuth\TestFramework\ControllerActionPredispatch` preference to be a plugin keeping the `\Magento\TwoFactorAuth\Observer\ControllerActionPredispatch` interceptor in play

### PR DESCRIPTION
**Pull request by fredden**
### Description
These preferences are defined in XML but the classes they reference do not exist (in production mode). This is blocking progress on magento/magento2#33161

### Fixed Issues
- Fixes #296

### Manual testing scenarios
No visible changes should be apparent with this change.

### Questions or comments
*None*

### Contribution checklist
 - [x] Author has signed the [Adobe CLA](https://opensource.adobe.com/cla.html)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: TwoFactorAuth/etc/di.xml [hunk 1]
```diff
@@ -40,7 +40,6 @@
     <preference for="Magento\TwoFactorAuth\Api\DuoAuthenticateInterface" type="Magento\TwoFactorAuth\Model\Provider\Engine\DuoSecurity\Authenticate"/>
     <preference for="Magento\TwoFactorAuth\Api\Data\DuoDataInterface" type="Magento\TwoFactorAuth\Model\Data\Provider\Engine\DuoSecurity\Data"/>
     <preference for="Magento\TwoFactorAuth\Api\U2fKeyConfigReaderInterface" type="Magento\TwoFactorAuth\Model\Provider\Engine\U2fKey\ConfigReader"/>
-    <preference for="OTPHP\TOTPInterface" type="OTPHP\TOTP"/>
```
Comment (nathanjosiah) [inline TwoFactorAuth/etc/di.xml:43]: Why was this removed?

**Review by nathanjosiah — CHANGES_REQUESTED** [review]

**Comment by nathanjosiah** [comment]
@fredden Thank you for your contribution! I think the PR looks ok except for a comment I left in the di.xml file. Please take a look at that and I think we can move this forward.

File: TwoFactorAuth/etc/di.xml (hunk 1 above)
Comment (fredden) [reply TwoFactorAuth/etc/di.xml:43]: This class isn't within scope for dependency injection.

Screenshot_2021-06-15_22-38-40

From what I can tell, this is used as a return type in `Magento\TwoFactorAuth\Model\Provider\Engine\Google::getTotp()` and `Magento\TwoFactorAuth\Model\Provider\Engine\Google\TotpFactory::create()`, and as a mock in `Magento\TwoFactorAuth\Test\Unit\Model\Provider\Engine\GoogleTest::setUp()`. None of these seem to need a Magento preference.

**Comment by fredden** [comment]
@magento run all tests

**Comment by nathanjosiah** [comment]
@magento run all tests with 2.4.3-develop

**Comment by nathanjosiah** [comment]
@magento run all tests against 2.4.3-develop

**Comment by nathanjosiah** [comment]
@fredden Looks like there are still some errors in the integration tests and the static tests.

**Comment by nathanjosiah** [comment]
@fredden If you want to work on this let me know, there are still valid failures in the builds which I have highlighted. I will close this for now.

**Comment by fredden** [comment]
@nathanjosiah Please can you re-open this. I don't seem to have a button to do so myself. This definitely needs to be fixed.

The error showing up that seems related to this change is:
> Data set: app/code/Magento/TwoFactorAuth/Plugin/BypassTwoFactorAuthForTestFramework.php
Module Magento\TwoFactorAuth has external dependencies: hard [Magento\TestFramework]
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/DependencyTest.php:572
/var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/DependencyTest.php:575

It doesn't feel right to put a runtime dependency on the test framework. The required class is distributed as part of `magento/magento2-base`. What's the best approach for resolving this?

I can see that the same test is reporting a number of similar issues in other modules. Perhaps that's something that should be resolved in a separate pull request?
> Module Magento\ReCaptchaCheckout has undeclared dependencies: hard [Magento\QuoteGraphQl]
Module Magento\ReCaptchaCustomer has undeclared dependencies: hard [Magento\CustomerGraphQl, Magento\Integration]
Module Magento\ReCaptchaNewsletter has undeclared dependencies: hard [Magento\NewsletterGraphQl]
Module Magento\ReCaptchaPaypal has undeclared dependencies: hard [Magento\PaypalGraphQl]
Module Magento\ReCaptchaReview has undeclared dependencies: hard [Magento\ReviewGraphQl]
Module Magento\ReCaptchaSendFriend has undeclared dependencies: hard [Magento\SendFriendGraphQl]

**Comment by nathanjosiah** [comment]
@fredden all of those errors are false-positives. There are no undocumented hard dependencies, if you look at the code for those there are runtime checks for classes to see if they are available first. It's a known issue but one we can safely ignore. You are welcome to continue to contribute to this issue, let's proceed by pulling in the latest develop and run all the tests again. The correct mainline version should be used by default now.

**Comment by fredden** [comment]
It looks like there are more out-of-scope test failures. It's hard to tell if there are any genuine test failures due to the changes within this pull request. How should we proceed?

**Comment by nathanjosiah** [comment]
@magento run all tests against 2.4.3

**Comment by nathanjosiah** [comment]
@magento run all tests against 2.4.4

**Comment by nathanjosiah** [comment]
@fredden Looks like everything is green except the static test (https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/security-package/pull/297/72113ff36dd11533a11b1852965b0149/Statics/allure-report-ce/index.html#categories/ace6c7944747e2a66397aa9d447b7430/fa91d33adabf7a98/) which doesn't like the hard dep in your new plugin.

```
app/code/Magento/TwoFactorAuth/Plugin/BypassTwoFactorAuthForTestFramework.php Module Magento\TwoFactorAuth has external dependencies: hard [Magento\TestFramework] /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/DependencyTest.php:572 /var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56 /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/DependencyTest.php:575
```

**Comment by fredden** [comment]
@magento run all tests against 2.4.4

**Comment by fredden** [comment]
I've changed the namespace for the plugin to see if this makes the static test happier. This doesn't seem to change the functionality any. (ie, this still fixes the bug in this module, and is compatible with the `setup:di:compile` check/safe-guard being introduced in the linked pull request.) I don't know if this will fix the static test or not; I suspect there's logic to ignore certain paths as the original Observer has exactly the same dependencies on the TestFramework.

I've triggered test runs for v2.4.4 (as @nathanjosiah has done previously) and the main branches (from the linked pull request so all the other related pull requests are included in the build).

**Comment by nathanjosiah** [comment]
The 2.4.4 tests are all green and the DB compare job is failing [here](https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/security-package/pull/297/05bc8a08a7e2e33251741380f7eab92f/DatabaseCompare/ee-sd-db-compare-report.html) and [here](https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/security-package/pull/297/05bc8a08a7e2e33251741380f7eab92f/DatabaseCompare/ee-skip-sd-db-compare-report.html) but they are both about the `magento_giftregistry_data` table which is clearly not from this PR. I am pulling this ticket in for QA and internal processing.

**Comment by nathanjosiah** [comment]
@magento export issue to Jira project MC as Bug

**Review by nathanjosiah — APPROVED** [review]

**Comment by nathanjosiah** [comment]
@magento export issue to Jira project MC as Bug

**Comment by nathanjosiah** [comment]
@magento import code to https://github.com/magento-commerce/security-package

**Comment by nathanjosiah** [comment]
@magento import code to https://github.com/magento-cia/security-package

**Comment by nathanjosiah** [comment]
@magento import code to https://github.com/magento-cia/security-package

**Comment by nathanjosiah** [comment]
@magento import code to https://github.com/magento-commerce/security-package

**Comment by magento-engcom-team** [comment]
@nathanjosiah, an error occurred during the Pull Request import.

**Comment by nathanjosiah** [comment]
@magento import code to https://github.com/magento-commerce/security-package

**Comment by magento-engcom-team** [comment]
@nathanjosiah the branch with code successfully imported into`magento-commerce/security-package` repository. Branch name: `imported-magento-security-package-297`.

**Comment by magento-engcom-team** [comment]
@nathanjosiah the branch with code successfully imported into`magento-commerce/security-package` repository. Branch name: `imported-magento-security-package-297`.
